kafka/topiccreator.go

```
@@ -0,0 +1,143 @@
+	TopicConfigs map[string]string
+}
+
+// Validate ensures the configuration is valid, returning an error otherwise.
+func (cfg TopicCreatorConfig) Validate() error {
+	var errs []error
+	if cfg.PartitionCount == 0 {
+		errs = append(errs, errors.New("kafka: partition count must be non-zero"))
+	}
+	return errors.Join(errs...)
+}
+
+// TopicCreator creates Kafka topics.
+type TopicCreator struct {
+	m                *Manager
+	partitionCount   int
+	origTopicConfigs map[string]string
+	topicConfigs     map[string]*string
+}
+
+// NewTopicCreator returns a new TopicCreator with the given config.
+func (m *Manager) NewTopicCreator(cfg TopicCreatorConfig) (*TopicCreator, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, fmt.Errorf("kafka: invalid topic creator config: %w", err)
+	}
+	var topicConfigs map[string]*string
+	if len(cfg.TopicConfigs) != 0 {
+		topicConfigs = make(map[string]*string, len(cfg.TopicConfigs))
+		for k, v := range cfg.TopicConfigs {
+			topicConfigs[k] = kadm.StringPtr(v)
+		}
+	}
+	return &TopicCreator{
+		m:                m,
+		partitionCount:   cfg.PartitionCount,
+		origTopicConfigs: cfg.TopicConfigs,
+		topicConfigs:     topicConfigs,
+	}, nil
+}
+
+// CreateTopics creates one or more topics.
+//
+// Topics that already exist will be left unmodified.
+func (c *TopicCreator) CreateTopics(ctx context.Context, topics ...apmqueue.Topic) error {
+	// TODO(axw) how should we record topics?
+	ctx, span := c.m.tracer.Start(ctx, "CreateTopics", trace.WithAttributes(
+		semconv.MessagingSystemKey.String("kafka"),
+	))
+	defer span.End()
+
+	topicNames := make([]string, len(topics))
+	for i, topic := range topics {
+		topicNames[i] = string(topic)
+	}
+	responses, err := c.m.client.CreateTopics(
+		ctx,
+		int32(c.partitionCount),
+		-1, // default.replication.factor
+		c.topicConfigs,
+		topicNames...,
+	)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, "CreateTopics returned an error")
```

Why not report the actual error?

```suggestion
		span.SetStatus(codes.Error, err.Error())
```

---

SGTM. I don't think we do anything with the status message at the moment, but perhaps we should.